For #26172 - New CFR popup composable

Mugurell opened the pull request:
This will allow for pinpoint accuracy when anchoring and resolve any color
disparities between the popup body and the indicator arrow by having everything
drawn programmatically as one shape.

Because of the async nature of the values for insets and screen rotation
immediately after an orientation change the popup will automatically get
dismissed to prevent any anchoring issues.
While not ideal the effect of this is better than accepting layout issues after
orientation changes and is the same approach used for other of our popups.


As an example of how easy it is to control anchoring programatically:

https://user-images.githubusercontent.com/11428869/153623802-6f09556f-86f5-4aa3-bdec-1f349f8c7493.mp4

The typography to be used is still in discussion - https://mozilla-hub.atlassian.net/browse/FNXV2-20991?focusedCommentId=572965

### Pull Request checklist
<!-- Before submitting the PR, please address each item -->
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA
<!-- Before submitting the PR, please address each item -->
- [ ] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.







### GitHub Automation
Fixes #26172

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+ * This is compiled to the underlying `Int` type so incurs no performance penalty.
+ */
+@JvmInline
+internal value class Pixels(val value: Int)
```
this could potentially be useful in another component. What if we broke this out into a separate PR where we add `.px` extension functions, similar to the `Dp` value class?

Mugurell replied:
I like the idea.
Fully supporting `Px` would mean adding many of the methods from the [`Dp.kt` file](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-unit/src/commonMain/kotlin/androidx/compose/ui/unit/Dp.kt) maybe to [AC around DisplayMetrics](https://github.com/mozilla-mobile/android-components/blob/71e124a5dbafffb800168ea615fdbc54d7398f50/components/support/ktx/src/main/java/mozilla/components/support/ktx/android/util/DisplayMetrics.kt) ?
Given the bigger scope I'd leave this as a later refactoring.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+    val dismissOnBackPress: Boolean = true,
+    val dismissOnClickOutside: Boolean = true,
+    val overlapAnchor: Boolean = false,
+    val indicatorArrowStartOffset: Dp = 30.dp,
+    val indicatorArrowHeight: Dp = 10.dp,
+    val cornerRadius: Dp = 10.dp,
+    val elevation: Dp = 10.dp,
```
where are the default specs for this obtained from?

Mugurell replied:
Thank you for raising this.
I just eyeballed the values to get everything working but did not came back to polish.
Checked now in previous CFRs  ([onboarding_jump_back_in_cfr](https://github.com/mozilla-mobile/fenix/blob/263e4d819be3b67c037917237f55050f3c09b8c9/app/src/main/res/layout/onboarding_jump_back_in_cfr.xml) and [onboarding_inactive_tabs_cfr](https://github.com/mozilla-mobile/fenix/blob/263e4d819be3b67c037917237f55050f3c09b8c9/app/src/main/res/layout/onboarding_inactive_tabs_cfr.xml)) and saw that they use [`15` as the "indicator offset"](https://github.com/mozilla-mobile/fenix/blob/263e4d819be3b67c037917237f55050f3c09b8c9/app/src/main/res/layout/onboarding_jump_back_in_cfr.xml#L55) and [`12` as the cornerRadius](https://github.com/mozilla-mobile/fenix/blob/51715c27b9f84a42b65be14ba3a0bf1cb9864e39/app/src/main/res/drawable/onboarding_popup_shape.xml#L6). Will update accordingly.
The other two properties - `indicatorArrowHeight` and `elevation` remain to be validated by UX maybe when I'll have the TCP CFR PR ready which would use this new CFR.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+        )
+
+        Popup(
+            popupPositionProvider = object : PopupPositionProvider {
+                override fun calculatePosition(
+                    anchorBounds: IntRect,
+                    windowSize: IntSize,
+                    layoutDirection: LayoutDirection,
+                    popupContentSize: IntSize
+                ): IntOffset {
+                    // Popup will be anchored such that the indicator arrow will point to the middle of the anchor View
+                    // but the popup is allowed some space as start padding in which it can be displayed such that the
+                    // indicator arrow is exactly at the top-start/bottom-start corner but slightly translated to end.
+                    // Values are in pixels.
+                    return IntOffset(
+                        when (layoutDirection) {
+                            LayoutDirection.Ltr -> popupBounds.startCoord.value
+                            else -> popupBounds.endCoord.value
+                        },
+                        when (properties.indicatorDirection) {
+                            IndicatorDirection.UP -> {
+                                when (properties.overlapAnchor) {
+                                    true -> anchorLocation.last() + anchor.height / 2
+                                    else -> anchorLocation.last() + anchor.height
+                                }
+                            }
+                            IndicatorDirection.DOWN -> {
+                                when (properties.overlapAnchor) {
+                                    true -> anchorLocation.last() - popupContentSize.height + anchor.height / 2
+                                    else -> anchorLocation.last() - popupContentSize.height
+                                }
+                            }
+                        }
+                    )
+                }
+            },
```
This takes up a considerable size of the instantiation for the `Popup`. Would it make sense to break this into out into a private function so the instantiation is a little cleaner?

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+ */
+@SuppressLint("ViewConstructor") // Intended to be used only in code, don't need a View constructor
+@VisibleForTesting
+internal class CFRPopupFullScreenLayout(
```
could you add the kdocs for the params here?

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+                text = text,
+                indicatorDirection = properties.indicatorDirection,
+                indicatorArrowStartOffset = with(LocalDensity.current) {
+                    (indicatorOffset.value).toDp()
```
```suggestion
                    indicatorOffset.value.toDp()
```

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+                    dismiss()
+                    onDismiss(true)
+                },
+                action = action
```
```suggestion
                action = action,
```

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+ */
+@SuppressLint("ViewConstructor") // Intended to be used only in code, don't need a View constructor
+@VisibleForTesting
+internal class CFRPopupFullScreenLayout(
```
This file is pretty large. Would it make sense to have `CFRPopupFullScreenLayout` and `CFRPopupContent` in their own files?

Mugurell replied:
Also thought about this but kept it here to better encapsulate it as it should not be used by other classes.
But the same can be said about `CFRPopupShape`.
I'll extract it to a new file, thanks!

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+    ): PopupHorizontalBounds {
+        val arrowIndicatorHalfWidth = arrowIndicatorWidth.value / 2
+
+        if (LocalConfiguration.current.layoutDirection == View.LAYOUT_DIRECTION_LTR) {
```
nit: the `return` could be inlined with the `if`

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+        ?: 0.coerceAtLeast(0)
+
+    @Px
+    private fun Dp.toPx(): Int {
+        return this.value
+            .dpToPx(anchor.resources.displayMetrics)
+            .roundToInt()
+    }
```
why use this over `fun Dp.roundToPx()` in `Density`?

Mugurell replied:
This allows to just write `dpValue.toPx()` to get what I want
vs having to write
`LocalDensity.current.run { dpValue.roundToPx() }`

Maybe another something to extract for later usages.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+ * Complete content of the popup.
+ * [CFRPopupShape] with a gradient background containing [text] and a dismiss ("X") button.
+ *
+ * @param text String offering more context about a particular feature to the user.
```
```suggestion
 * @param text String message in the popup.
```

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+    cornerRadius: Dp,
+    elevation: Dp = 0.dp,
+    onDismiss: () -> Unit,
+    action: @Composable (() -> Unit)? = null
```
where do we foresee this being used?

Mugurell replied:
Guess that if we have a constants width of the CFR popups the `indicatorArrowHeight` and `cornerRadius` should also be constants. Will update as such. Will update as such.

`onDismiss` will be used for side effects like telemetry
`action` can be another composable for a button / link.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+    cornerRadius: Dp,
+    elevation: Dp = 0.dp,
+    onDismiss: () -> Unit,
+    action: @Composable (() -> Unit)? = null
```
```suggestion
    action: @Composable (() -> Unit) = {}
```

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+                        style = FirefoxTheme.typography.body2
+                    )
+
+                    action?.invoke()
```
```suggestion
                    action()

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+    onDismiss: () -> Unit,
+    action: @Composable (() -> Unit)? = null
+) {
+    FirefoxTheme {
```
this should be moved to `Content()` in `CFRPopupFullScreenLayout`

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+                color = Color.Transparent,
+                elevation = elevation,
+                // Need to override the default RectangleShape to avoid casting shadows for that shape.
+                shape = CFRPopupShape(
+                    indicatorDirection, indicatorArrowStartOffset, indicatorArrowHeight, cornerRadius
+                ),
```
if this is the same shape used in the modifier, it could be initialized once in a local `val` (and with the named params)

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,610 @@
+    action: @Composable (() -> Unit)? = null
+) {
+    FirefoxTheme {
+        Box(modifier = Modifier.width(POPUP_WIDTH_DP.dp + HORIZONTAL_PADDING.dp * 2)) {
+            Surface(
```
A `Surface` is a wrapper for a `Box`, so the two containers could be merged here by just having the `.width` Modifier before `.align`

Mugurell replied:
[`align` is scoped to Box](https://cs.android.com/androidx/platform/tools/dokka-devsite-plugin/+/master:testData/compose/source/androidx/compose/foundation/layout/Box.kt;l=219?q=interface%20BoxScope) and I need that to position the `X` button.
Then I use the Surface because it knows to easily handle a shape plus elevation.

MozillaNoah replied:
Oh dang ok, fair enough. Weird that they didn't trickle down the functionality to `Surface`.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopupShape.kt:
```diff
@@ -0,0 +1,252 @@
+}
+
+@Composable
+@Preview(locale = "en", name = "LTR")
+@Preview(locale = "ar", name = "RTL")
```
nit: do we need additional previews for the Shape for RTL and LTR?

Mugurell replied:
I do think so. Would sort of us these previews as screenshot tests.
Given that the shapes are drawn programatically and there is different code to draw LRT and RTL these previews do help for a quick validation.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopupShape.kt:
```diff
@@ -0,0 +1,252 @@
+import org.mozilla.fenix.theme.FirefoxTheme
+import kotlin.math.roundToInt
+
+private const val INDICATOR_BASE_TO_HEIGHT_RATIO = 1.5f
```
what's a TLDR for this const?

Mugurell replied:
A value of `1` would mean the indicator is a equilateral triangle. 
A different value would give an isosceles triangle.
This would be useful in supporting different heights of the indicator with a big enough base. But will probably standardize the look based on UX feedback and remove this.

MozillaNoah reviewed (commented):
Looking good so far

MozillaNoah commented:
We may also want a UX green light for this

gabrielluong commented on app/src/main/res/values/strings.xml:
```diff
@@ -73,6 +73,9 @@
     <!-- Text for the negative action button -->
     <string name="open_in_app_cfr_negative_button_text">Dismiss</string>
 
+    <!-- Content description for close button used of "contextual feature recommendation" (CFR) popups -->
```
```suggestion
    <!-- Content description for close button used in "contextual feature recommendation" (CFR) popups -->
```

Mugurell commented:
Thank you for the (quick) help!
The plan is to have this is a landable form, use it for the TCP CFR and provide an apk for UX to help with the final design validation before merging.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopupContent.kt:
```diff
@@ -0,0 +1,176 @@
+                .size(48.dp)
+        ) {
+            Icon(
+                Filled.Close,
```
is this missing a named parameter?

MozillaNoah reviewed (commented):
There's a weird box being drawn in `CFRPopupContent` that's only showing up in the previews with a downward indicator triangle.

![image](https://user-images.githubusercontent.com/87384386/181064178-c47f6eb6-7fff-41be-a9ab-980644164d34.png)

Mugurell commented:
> There's a weird box being drawn in `CFRPopupContent` that's only showing up in the previews with a downward indicator triangle.

Interesting, thanks. Removed setting the elevation in the `Surface` from `CFRPopupContent`. Will default to the elevation the compose Popup uses.

Mugurell commented:
Refactored to add the static properties declared in `CFRPoup` to a companion object to avoid polluting the general namespace and ensure a bit more context for where those values are used/useful.
Added on Jira - https://mozilla-hub.atlassian.net/browse/FNXV2-21106 more details for Kate to confirm or not.

MozillaNoah commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopupContent.kt:
```diff
@@ -0,0 +1,171 @@
+        ) {
+            Icon(
+                imageVector = Filled.Close,
+                contentDescription = stringResource(string.cfr_dismiss_button_default_content_description),
```
Was the `R.string` imported?
```suggestion
                contentDescription = stringResource(R.string.cfr_dismiss_button_default_content_description),
```

Mugurell replied:
Yes, thank you!
Move some code around and imports got mangled, also added some `R.`s, will check again for all.

MozillaNoah reviewed (approved):
This is good to go otherwise.

mavduevskiy commented on app/src/test/java/org/mozilla/fenix/compose/cfr/CFRPopupFullScreenLayoutTest.kt:
```diff
@@ -0,0 +1,100 @@
+@RunWith(FenixRobolectricTestRunner::class)
+class CFRPopupFullScreenLayoutTest {
+    @Test
+    fun `WHEN the popup is constructed THEN setup lifecycle owners`() {
```
I have been playing around with the tests, and have noticed that both `ViewTreeSavedStateRegistryOwner.get()` and `ViewTreeLifecycleOwner.get()` always return `null`.

Since they are static, do we have any options but to check if the have been called with proper arguments?
Something like this.

```
    @Test
    fun `WHEN the popup is constructed THEN setup lifecycle owners`() {
        mockkStatic(ViewTreeLifecycleOwner::class)
        mockkStatic(ViewTreeSavedStateRegistryOwner::class)

        val container = View(testContext)
        val popupView = CFRPopupFullScreenLayout("", container, mockk(), mockk()) {}

        val viewTreeLifecycleCaptor = slot<View>()
        val viewTreeSavedStateCaptor = slot<View>()

        verify { ViewTreeLifecycleOwner.set(capture(viewTreeLifecycleCaptor), any()) }
        verify { ViewTreeSavedStateRegistryOwner.set(capture(viewTreeSavedStateCaptor), any()) }

        assertEquals(viewTreeLifecycleCaptor.captured, popupView)
        assertEquals(viewTreeSavedStateCaptor.captured, popupView)
    }
```

Mugurell replied:
Thank you!
Agree that checking for `null` equality is not really useful.
Your proposed approach would work in checking that whatever value the above "..owner"s have are corrently set, even if they are null.
For a somewhat simpler approach and actually checking for a "real" instance though we can set "..owner"s for the anchor and then test that they are properly propagated to our popupView (after which they will be propagated to `androidx.compose.ui.window.PopupLayout`).

mavduevskiy commented on app/src/test/java/org/mozilla/fenix/compose/cfr/CFRPopupFullScreenLayoutTest.kt:
```diff
@@ -0,0 +1,100 @@
+
+        popupView.dismiss()
+
+        assertEquals(null, ViewTreeLifecycleOwner.get(popupView))
```
This check doesn't seem to fulfill its purpose since it is null before `.dismiss()` and `.show()` as well

mavduevskiy replied:
If we want to check if `ViewTreeLifecycleOwner.set(this, null)` was called `onDismiss()`, wouldn't mocking static methods and checking if it was called with `null` lifecycle do the trick?

Mugurell replied:
Thank you!
Same as above we can easily test even with a real initial owner that is actually set to null after the call to `dismiss`.

mavduevskiy reviewed (commented):
Looks like a lot of (awesome) work! 
There hasn't been much discussion on tests, so I played around it a bit. I hope I am being helpful.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopup.kt:
```diff
@@ -0,0 +1,123 @@
+         * How tall the indicator arrow should be.
+         * This will also affect how wide the base of the indicator arrow will be.
+         */
+        internal const val DEFAULT_INDICATOR_HEIGHT = 15
```
Should this be 16dp? https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/res/values/dimens.xml#45

Mugurell replied:
The arrow drawable that `cfr_triangle_height == 16dp` applies to - [`ic_cfr_triangle`](https://github.com/mozilla-mobile/fenix/blob/main/app/src/main/res/drawable/ic_cfr_triangle.xml) actually has a lot of padding so the visual height will be maybe half of that value, while `cfr_triangle_width == 16dp` applied to the same drawable would probably result visually in a width of ~15dp.

![image](https://user-images.githubusercontent.com/11428869/182680976-9d63f25b-c416-45ab-8df9-a5eeefd4c39b.png)

&nbsp;

To me the arrow seemed a bit small in height so I tried a higher default.
Kate just confirmed the designs shown in https://mozilla-hub.atlassian.net/browse/FNXV2-21106 look good so I would say we can go with the value of `15dp` from here and change afterwards if needed.

gabrielluong commented on app/src/main/java/org/mozilla/fenix/compose/cfr/CFRPopupContent.kt:
```diff
@@ -0,0 +1,171 @@
+@Composable
+@Preview(locale = "en", name = "LTR")
+@Preview(locale = "ar", name = "RTL")
+@Preview(uiMode = Configuration.UI_MODE_NIGHT_YES, name = "Night mode")
+@Preview(uiMode = Configuration.UI_MODE_NIGHT_NO, name = "Light mode")
```
```suggestion
@Preview(uiMode = Configuration.UI_MODE_NIGHT_YES, name = "Dark theme")
@Preview(uiMode = Configuration.UI_MODE_NIGHT_NO, name = "Light theme")
```
I would prefer if we could go with the current theme terminology within Fenix

gabrielluong reviewed (commented):
I find myself looking into this since I pointed out to @sarah541 that it might be worth using this CFR pop composable to show the CFR for the synced tab card in the Jump back in section.

mavduevskiy commented on app/src/test/java/org/mozilla/fenix/compose/cfr/CFRPopupFullScreenLayoutTest.kt:
```diff
@@ -0,0 +1,112 @@
+
+        popupView.dismiss()
+
+        assertNull(popupView.findViewTreeLifecycleOwner())
```
Does it make sense to check if it was not `null` before `dismiss` is called? Technically speaking, it could have been `null` all along :)

Mugurell replied:
My reasoning was that we have a test already just above for checking that upon instantiating the popupView the lifecycle owners will not be null.
I can add a check here also to be sure.

Mugurell commented:
Landing this for Sarah to also be able to use it.
